proposals/XXXX-improve-interaction-between-private-declarations-and-extensions.md

```diff
@@ -0,0 +1,121 @@
+
+The prevalence of `fileprivate` in practice has caused mixed reactions from Swift developers, culminating in proposal [SE-0159](0159-fix-private-access-levels.md) which suggested reverting the access control model to Swift 2’s design.  That proposal was rejected for two reasons: scoped access is something that many developers use and value, and because it was seen as too large of a change given Swift 4’s source compatibility goals.
+
+In contrast to SE-0159, this proposal is an extremely narrow change (which is almost completely additive) to the SE-0025 model, which embraces the extension oriented design of Swift.  The authors believe that this change will not preclude introduction of submodules in the future.
+
+
+## Detailed Design
+
+The only change this proposal introduces is for extensions of a type defined in the same source file as the type itself.  Simply, the access control scope of the extension is considered to be the access control scope of the extended type.
+
+This has two ramifications: 
+
+* Declarations in one of these extensions get access to the `private` members of the type.
+
+* If the declaration in the extension itself is defined as `private`, then they are accessible to declarations in the type, and other extensions of that type (in the same file).
+
+
+Here is a simple code example that demonstrates this:
+
+```swift
+struct S {
+    private var p: Int
+
+    func f() { 
+        use(g())    // ok, g() is accessible within S
+    }
+}
+
+extension S {
+    private func g() {
+        use(p)      // ok, g() has access to p, since it is in an extension on S.
+    }
+}
+
+extension S {
+    func h() {
+        use(g())    // Ok, h() has access to g() since it defined in the access control scope for S.
+    }
+}
+```
+
+Please note:
+
+* This visibility does **not** extend to subclasses of a class in the same file, it only affects extensions of the type itself.
+
+* Constrained extensions are extensions, so this visibility **does** extend to them as well.  For example, the body of `extension Optional where Wrapped == String {  }` would have access to `private` members of Optional, assuming the extension is defined in the same file as `Optional`.
+
+* This proposal does not change access control behavior for types nested within each other.  As in Swift 3, inner types have access to the private members of outer types, but outer types cannot refer to private members of inner types.  For example:
+
+```swift
+struct Outer {
+    private var outerValue = 42
+    
+    struct Inner {
+        private var innerValue = 57
+
+        func innerTest() {
+            print(outerValue)    // still ok.
+        }
+    }
+
+    func test(_ i: Inner) {
+        print(i.innerValue)      // still an error
+    }
+}
+```
+
```

Comment:
I think it can't hurt because even if the language in the introduction and design is clear, its a case that people won't necessarily think about.